qiskit/circuit/quantumcircuit.py

```diff
@@ -106,113 +106,859 @@
+            as a whole.
+
+        :ref:`circuit-adding-data-objects`
+            The methods for adding new :class:`~.expr.Var` variables to a circuit after
+            initialization.
+
+    .. automethod:: get_var
+    .. automethod:: has_var
+    .. automethod:: iter_captured_vars
+    .. automethod:: iter_declared_vars
+    .. automethod:: iter_input_vars
+    .. automethod:: iter_vars
+
+
+    .. _circuit-adding-operations:
+
+    Adding operations to circuits
+    =============================
+
+    You can add anything that implements the :class:`.Operation` interface to a circuit as a single
+    instruction, though most things you will want to add will be :class:`~.circuit.Instruction` or
+    :class:`~.circuit.Gate` instances.
+
+    .. seealso::
+        :ref:`circuit-operations-instructions`
+            The :mod:`qiskit.circuit`-level documentation on the different interfaces that Qiskit
+            uses to define circuit-level instructions.
+
+    .. _circuit-append-compose:
+
+    Driver methods to add operations
+    --------------------------------
+
+    These are the base methods that handle adding any object, including user-defined ones, onto
+    circuits.
+
+    ===============  ===============================================================================
+    Method           When to use it
+    ===============  ===============================================================================
+    :meth:`append`   Add an instruction as a single object onto a circuit.
+    :meth:`_append`  Same as :meth:`append`, but a low-level interface that elides almost all error
+                     checking.
+    :meth:`compose`  Inline the instructions from one circuit onto another.
+    :meth:`tensor`   Like :meth:`compose`, but strictly for joining circuits that act on disjoint
+                     qubits.
+    ===============  ===============================================================================
+
+    :class:`QuantumCircuit` has two main ways that you will add more operations onto a circuit.
+    Which to use depends on whether you want to add your object as a single "instruction"
+    (:meth:`append`), or whether you want to join the instructions from two circuits together
+    (:meth:`compose`).
+
+    A single instruction or operation appears as a single entry in the :attr:`data` of the circuit,
+    and as a single box when drawn in the circuit visualizers (see :meth:`draw`).  A single
+    instruction is the "unit" that a hardware backend might be defined in terms of (see
+    :class:`.Target`).  An :class:`~.circuit.Instruction` can come with a
+    :attr:`~.circuit.Instruction.definition`, which is one rule the transpiler (see
+    :mod:`qiskit.transpiler`) will be able to fall back on to decompose it for hardware, if needed.
+    An :class:`.Operation` can only be decomposed if it has some associated high-level synthesis
+    method registered for it (see :mod:`qiskit.transpiler.passes.synthesis.plugin`).
```

Comment:
I think we should be a bit more clear here, because an `Instruction` and `Gate` object are also `Operation`s. There is also the complexity around opaque instructions and gates which don't include a definition. Which is covered by the previous sentence but not really expanded upon.

```suggestion
    A direct implementer of the :class:`.Operation` interface can only be decomposed if it has some associated high-level synthesis
    method registered for it (see :mod:`qiskit.transpiler.passes.synthesis.plugin`) as it does not include a default
    decomposition via :attr`.Instruction.definition`.
```

Reply:
Done in 956475957f.